spec/elastic_apm/transport/filters_spec.rb

```diff
@@ -44,6 +44,19 @@ module Transport
           expect(untouched).to_not have_received(:call)
         end
       end
+
+      describe 'thread safety' do
+        it 'may add filters while applying' do
+          threads =
+            (0...100).map do |i|
+              Thread.new { subject.apply!(payload: i) }
+            end
+
+          subject.add :new, ->(_) { '' }
```

Comment:
Were you able to reproduce the bug with this test before adding a fix? I ran this without the fix in `Filters` and couldn't get it to fail.

This only gives the `add` and `apply` one chance to step on each other. I would suggest instead increasing the likelihood of the error by adding many new keys into the hash while applying the filters.
I'm able to reproduce the bug every time I run this test:

```ruby
    context 'altering and accessing filters at the same time' do
        it 'does not mutate the original' do
          threads = []
          100.times do |i|
            threads <<
              if i % 2 == 0
                Thread.new { subject.apply!(payload: i) }
              else
                Thread.new { subject.add "filter#{i}", ->(_) {''} }
              end
          end
          threads.each(&:join)
        end
      end
```

```ruby
$ rspec spec/elastic_apm/transport/filters_spec.rb:48
Randomized with seed 29514

ElasticAPM::Transport::Filters
  altering and accessing filters at the same time
    does not mutate the original (FAILED - 1)

Failures:

  1) ElasticAPM::Transport::Filters altering and accessing filters at the same time does not mutate the original
     Failure/Error: @filters[key] = filter
     
     RuntimeError:
       can't add a new key into hash during iteration
     # ./lib/elastic_apm/transport/filters.rb:22:in `add'
     # ./spec/elastic_apm/transport/filters_spec.rb:56:in `block in Transport'
```